File: ROS_YuXiang/test5.2_tf_python/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer
from tf_transformations import quaternion_from_euler, euler_from_quaternion


class TfListener(Node):
    def __init__(self):
        super().__init__("tf2_listener")
        self.buffer_ = Buffer()
        self.listener_ = TransformListener(self.buffer_, self)
        self.timer_ = self.create_timer(1, self.get_transform)

    def get_transform(self):
        try:
            result = self.buffer_.lookup_transform(
                "base_link", "bottle_link", rclpy.time.Time(seconds=0),rclpy.time.Duration(seconds=1))
            transform = result.transform
            q = euler_from_quaternion([
                transform.rotation.x,
                transform.rotation.y,
                transform.rotation.z,
                transform.rotation.w
            ])
            self.get_logger().info(
                (f"transform.translation: {transform.translation},"
                 "rotation_quater: {q}, rotation_eular: {transform.rotation}"))
        except Exception as e:
            self.get_logger().warn(f"can not receive transform, due to {str(e)}")


def main():
    rclpy.init()
    node = TfListener()
    rclpy.spin(node)
    rclpy.shutdown()

File: ROS_YuXiang/test5.2_tf_python/src/demo_cpp_tf/src/tf_listener.cpp
#include <memory>
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/utils.h"
#include <chrono>

using namespace std::chrono_literals;

class TfListener : public rclcpp::Node
{
public:
    TfListener() : Node("tf_listener")
    {
        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, this);
        timer_ = this->create_wall_timer(5s, std::bind(&TfListener::getTransform, this));
    }

    void getTransform()
    {
        try
        {
            const auto transform = tf_buffer_->lookupTransform("base_link", "target_point",
                                                               this->get_clock()->now(), rclcpp::Duration::from_seconds(1.0f));
            const auto &translation = transform.transform.translation;
            const auto &rotation = transform.transform.rotation;
            double yaw, pitch, roll;
            tf2::getEulerYPR(rotation, yaw, pitch, roll);
            RCLCPP_INFO(this->get_logger(), "Translation: x: %f, y: %f, z: %f", translation.x, translation.y, translation.z);
            RCLCPP_INFO(this->get_logger(), "Rotation: yaw: %f, pitch: %f, roll: %f", yaw, pitch, roll);
        }
        catch (tf2::TransformException &ex)
        {
            RCLCPP_ERROR(this->get_logger(), "Could not transform: %s", ex.what());
        }
    }

private:
    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TfListener>());
    rclcpp::shutdown();
    return 0;
}
